### PR TITLE
Diagnostic service cleanup

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Cleanup that has come out of current LSP work.  Main areas are:

1. Use immutable arrays in IDiagnosticService.
2. Rename a type from DiagnsoticUpdateArgs to DiagnosticBucket for clarity.
3. Add helpers for common diagnostic service requests.
4. Make the error list scaffolding code strongly typed.

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs [hunk 1]
```diff
@@ -152,28 +152,28 @@ private void ProduceTags(TaggerContext<TTag> context, DocumentSnapshotSpan spanT
             var suppressedDiagnosticsSpans = (NormalizedSnapshotSpanCollection?)null;
             buffer?.Properties.TryGetProperty(PredefinedPreviewTaggerKeys.SuppressDiagnosticsSpansKey, out suppressedDiagnosticsSpans);
 
-            var eventArgs = _diagnosticService.GetDiagnosticsUpdatedEventArgs(
+            var buckets = _diagnosticService.GetDiagnosticBuckets(
                 workspace, document.Project.Id, document.Id, context.CancellationToken);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs:156]: this was always super oddly named.  nothing is being named.  nothing is an event.  nothing is an arg.  This was very confusing when doing the LSP changes.  'bucket' seems like a more reasonable name in that given any Doc in the IDIagnosticsService, there are N buckets that have diagnostics associated with them that are aggregated to form the full set of diagnostics for that document.

File: src/EditorFeatures/Test/Diagnostics/MockDiagnosticService.cs [hunk 2]
```diff
@@ -34,34 +34,22 @@ public MockDiagnosticService()
         {
         }
 
-        public IEnumerable<DiagnosticData> GetDiagnostics(Workspace workspace, ProjectId projectId, DocumentId documentId, object id, bool includeSuppressedDiagnostics, CancellationToken cancellationToken)
+        public ImmutableArray<DiagnosticData> GetDiagnostics(Workspace workspace, ProjectId? projectId, DocumentId? documentId, object? id, bool includeSuppressedDiagnostics, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test/Diagnostics/MockDiagnosticService.cs:37]: NRT'ed a few apis as well.

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioBaseDiagnosticListTable.LiveTableDataSource.cs [hunk 3]
```diff
@@ -35,7 +35,7 @@ internal abstract partial class VisualStudioBaseDiagnosticListTable
         /// See <see cref="VisualStudioDiagnosticListTableWorkspaceEventListener.VisualStudioDiagnosticListTable.BuildTableDataSource"/>
         /// for error list diagnostic source for "Build only" setting.
         /// </summary>
-        protected class LiveTableDataSource : AbstractRoslynTableDataSource<DiagnosticTableItem>
+        protected class LiveTableDataSource : AbstractRoslynTableDataSource<DiagnosticTableItem, DiagnosticsUpdatedArgs>
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioBaseDiagnosticListTable.LiveTableDataSource.cs:38]: `DiagnosticsUpdatedArgs` is still a wonky name.  can maybe remove that in the future as well.

File: src/Features/Core/Portable/Diagnostics/DiagnosticService.cs [hunk 4]
```diff
@@ -214,108 +215,95 @@ private void OnCleared(object sender, EventArgs e)
...
-
-                AppendMatchingData(source, workspace, projectId, documentId, null, list.Object);
-
-                foreach (var data in list.Object)
-                {
-                    yield return new UpdatedEventArgs(data.Id, data.Workspace, data.ProjectId, data.DocumentId);
-                }
+                AppendMatchingData(source, workspace, projectId, documentId, id: null, buffer);
+                foreach (var data in buffer)
+                    result.AddRange(new DiagnosticBucket(data.Id, data.Workspace, data.ProjectId, data.DocumentId));
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/DiagnosticService.cs:299]:
```suggestion
                    result.Add(new DiagnosticBucket(data.Id, data.Workspace, data.ProjectId, data.DocumentId));
```

File: src/Features/Core/Portable/Diagnostics/IDiagnosticService.cs [hunk 5]
```diff
@@ -25,11 +26,12 @@ internal interface IDiagnosticService
         /// <summary>
         /// Get current diagnostics stored in IDiagnosticUpdateSource
         /// </summary>
-        IEnumerable<DiagnosticData> GetDiagnostics(Workspace workspace, ProjectId projectId, DocumentId documentId, object id, bool includeSuppressedDiagnostics, CancellationToken cancellationToken);
+        ImmutableArray<DiagnosticData> GetDiagnostics(Workspace workspace, ProjectId? projectId, DocumentId? documentId, object? id, bool includeSuppressedDiagnostics, CancellationToken cancellationToken);
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/IDiagnosticService.cs:29]: We should make sure none of the internal IVT clients use the changed APIs, or we might have to do a co-ordinated insertion.

Comment (mavasani) [reply src/Features/Core/Portable/Diagnostics/IDiagnosticService.cs:29]: Actually, given just the return type is changing, we should probably be just fine.

**Review by mavasani — APPROVED** [review]
Thanks for the cleanup!

**Comment by mavasani** [comment]
Seems like bunch of diagnostics tests are failing.

File: src/Features/Core/Portable/Diagnostics/DiagnosticService.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Diagnostics/DiagnosticService.cs:299]: indeed.

**Pull request merged**